presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSource.java

```diff
@@ -0,0 +1,254 @@
+import static java.lang.Float.floatToRawIntBits;
+import static java.lang.Float.parseFloat;
+import static java.lang.Long.parseLong;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class IcebergPageSource
+        implements ConnectorPageSource
+{
+    private static final int MAX_BATCH_SIZE = 8 * 1024;
+
+    private final Block[] prefilledBlocks;
+    private final int[] delegateIndexes;
+    private final ConnectorPageSource delegate;
+
+    public IcebergPageSource(
+            List<IcebergColumnHandle> columns,
+            Map<Integer, IcebergPartitionKey> partitionKeysById,
+            ConnectorPageSource delegate,
+            TimeZoneKey timeZoneKey)
+    {
+        int size = requireNonNull(columns, "columns is null").size();
+        requireNonNull(partitionKeysById, "partitionKeysById is null");
+        this.delegate = requireNonNull(delegate, "delegate is null");
+
+        this.prefilledBlocks = new Block[size];
+        this.delegateIndexes = new int[size];
+
+        int outputIndex = 0;
+        int delegateIndex = 0;
+        for (IcebergColumnHandle column : columns) {
+            if (partitionKeysById.containsKey(column.getId())) {
+                IcebergPartitionKey partitionKey = partitionKeysById.get(column.getId());
+                Type type = column.getType();
+                Object prefilledValue = deserializePartitionValue(type, partitionKey.getValue(), column.getName(), timeZoneKey);
+                prefilledBlocks[outputIndex] = RunLengthEncodedBlock.create(column.getType(), prefilledValue, MAX_BATCH_SIZE);
+                delegateIndexes[outputIndex] = -1;
+            }
+            else {
+                delegateIndexes[outputIndex] = delegateIndex;
+                delegateIndex++;
+            }
+            outputIndex++;
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegate.getCompletedBytes();
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return delegate.getReadTimeNanos();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return delegate.isFinished();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        try {
+            Page dataPage = delegate.getNextPage();
+            if (dataPage == null) {
+                return null;
+            }
+            int batchSize = dataPage.getPositionCount();
+            Block[] blocks = new Block[prefilledBlocks.length];
+            for (int i = 0; i < prefilledBlocks.length; i++) {
+                if (prefilledBlocks[i] != null) {
+                    blocks[i] = prefilledBlocks[i].getRegion(0, batchSize);
+                }
+                else {
+                    blocks[i] = dataPage.getBlock(delegateIndexes[i]);
+                }
+            }
+            return new Page(batchSize, blocks);
+        }
+        catch (PrestoException e) {
```

Comment:
This can simplify to
```java
catch (RuntimeException e) {
    closeWithSuppression(e);
    throwIfInstanceOf(e, PrestoException.class);
    throw new PrestoException(ICEBERG_BAD_DATA, e);
}